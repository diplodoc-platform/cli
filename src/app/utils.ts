import {TextSizes, Theme} from '@doc-tools/components';

const DEFAULT_USER_SETTINGS = {
    theme: Theme.Light,
    textSize: TextSizes.m,
    showMiniToc: true,
    wideFormat: true,
    fullScreen: false,
};

export function getDocSettings() {
    const {
        theme: defaultTheme,
        textSize: defaultTextSize,
        showMiniToc: defaultShowMiniToc,
        wideFormat: defaultWideFormat,
        fullScreen: defaultFullScreen,
    } = DEFAULT_USER_SETTINGS;

    const theme = getSetting('theme') as Theme || defaultTheme;
    const textSize = getSetting('textSize') as TextSizes || defaultTextSize;
    const showMiniToc = getSetting('showMiniToc') || defaultShowMiniToc;
    const wideFormat = getSetting('wideFormat') || defaultWideFormat;
    const fullScreen = getSetting('fullScreen') || defaultFullScreen;

    return {
        theme,
        textSize,
        showMiniToc: strToBoolean(showMiniToc),
        wideFormat: strToBoolean(wideFormat),
        fullScreen: strToBoolean(fullScreen),
    };
}

export function getSetting(name: string) {
    return sessionStorage.getItem(name);
}

export function saveSetting<T>(name: string, value: T) {
    sessionStorage.setItem(name, String(value));
}

export function strToBoolean(str: any) {
    if (typeof str === 'boolean') {
        return str;
    }

    return str ? str === 'true' : false;
}

export function withSavingSetting<T>(settingName: string, onChange: (value: T) => void) {
    return (value: T) => {
        saveSetting<T>(settingName, value);

        onChange(value);
    };
}

export function updateRootClassName(theme: Theme, isMobile = false) {
    const themeClassName = theme === 'light' ? 'yc-root_theme_light' : 'yc-root_theme_dark';
    const mobileClassName = isMobile ? 'mobile' : '';

    document.body.className = `yc-root ${themeClassName} ${mobileClassName}`;
}
