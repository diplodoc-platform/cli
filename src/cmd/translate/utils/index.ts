import {ok} from 'assert';
import {basename, dirname, extname, resolve} from 'path';
import {readFileSync} from 'node:fs';
import glob from 'glob';

export {dumpFile, loadFile, resolveSchemas} from './fs';
export {extract, compose} from './translate';
export {TranslateError, LimitExceed, RequestError, AuthError} from './errors';

type TranslateArgs = {
    input: string;
    output?: string;
    source?: string;
    sourceLanguage?: string;
    sourceLanguageLocale?: string;
    target?: string;
    targetLanguage?: string;
    targetLanguageLocale?: string;
    auth?: string;
    folder?: string;
    glossary?: string;
    include?: string[] | string;
    exclude?: string[] | string;
    dryRun?: boolean;
    useSource?: boolean;
} & {
    [prop: string]: any;
};

export type TranslateParams = {
    input: string;
    output: string;
    source: [string, string];
    targets: [string, string][];
    auth?: string;
    folder?: string;
    glossary?: string;
    files: string[];
    dryRun: boolean;
    useSource: boolean;
};

export function normalizeParams(params: TranslateArgs, exts = EXTS): TranslateParams {
    const source = normalizeLocale(params, 'source')[0];
    const targets = normalizeLocale(params, 'target');
    const {input, files} = normalizeInput(params, source[0], exts);
    const {output, auth, folder, dryRun, useSource} = params;

    ok(input, 'Required param input is not configured');

    return {
        input,
        output: output || input,
        auth,
        folder,
        source,
        targets,
        files: [...new Set(files)],
        dryRun: Boolean(dryRun),
        useSource: Boolean(useSource),
    };
}

function normalizeLocale(params: TranslateArgs, scope: 'source' | 'target'): [string, string][] {
    const _scopeLanguage = params[scope + 'Language'];
    const _scopeLanguageLocale = params[scope + 'LanguageLocale'];
    const _scope = params[scope] || _scopeLanguageLocale || _scopeLanguage;

    if (!_scope) {
        return [['', '']];
    }

    if (Array.isArray(_scope)) {
        return _scope.map((_scope) => _scope.split('-'));
    }

    return [_scope.split('-') as [string, string]];
}

const EXTS = ['.md', '.yaml', '.json'];

function normalizeInput(params: TranslateArgs, language: string, exts: string[]) {
    let {input, include = [], exclude = []} = params;

    include = ([] as string[]).concat(include);
    exclude = ([] as string[]).concat(exclude);

    let files: string[] | null = null;
    if (extname(input) === '.list') {
        const list = readFileSync(input, 'utf8').split('\n');
        input = dirname(input);
        files = list.map((file) => {
            const absPath = resolve(input, file);

            if (!absPath.startsWith(input)) {
                throw new Error(`Insecure access to file out of project scope. (file: ${absPath})`);
            }

            if (!exts.includes(extname(file))) {
                throw new Error(`Unhandles file extension. (file: ${absPath})`);
            }

            return absPath;
        });
    } else {
        if (exts.includes(extname(input))) {
            files = [basename(input)];
            input = dirname(input);
        }

        if (!include.length) {
            include.push('...');
        }

        include = include.reduce((acc, item) => {
            if (item === '...') {
                acc.push(...exts.map((ext) => (language || '.') + '/**/*' + ext));
            } else {
                acc.push(item);
            }

            return acc;
        }, [] as string[]);

        files =
            files ||
            ([] as string[]).concat(
                ...include.map((match) =>
                    glob.sync(match, {
                        cwd: input,
                        ignore: exclude,
                        nodir: true,
                    }),
                ),
            );
    }

    return {input, files};
}

export class Defer<T = string> {
    resolve!: (text: T) => void;

    reject!: (error: any) => void;

    promise: Promise<T>;

    constructor() {
        this.promise = new Promise((resolve, reject) => {
            this.resolve = resolve;
            this.reject = reject;
        });
    }
}

export function bytes(texts: string[]) {
    return texts.reduce((sum, text) => sum + text.length, 0);
}
