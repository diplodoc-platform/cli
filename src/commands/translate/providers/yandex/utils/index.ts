export {LimitExceed, RequestError, AuthError} from './errors';

export class Defer<T = string> {
    resolve!: (text: T) => void;

    reject!: (error: any) => void;

    promise: Promise<T>;

    constructor() {
        this.promise = new Promise((resolve, reject) => {
            this.resolve = resolve;
            this.reject = reject;
        });
    }
}

export function bytes(texts: string[]) {
    return texts.reduce((sum, text) => sum + text.length, 0);
}
