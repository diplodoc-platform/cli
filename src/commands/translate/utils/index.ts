export type {Locale} from './config';
export {resolveSchemas, FileLoader} from './fs';
export {extract, compose} from './translate';
export {resolveSource, resolveTargets, resolveFiles, resolveVars} from './config';
export {
    TranslateError,
    ExtractError,
    ComposeError,
    SkipTranslation,
    EmptyTokensError,
} from './errors';
