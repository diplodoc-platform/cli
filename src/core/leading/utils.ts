import {cloneDeepWith, isString} from 'lodash';

export function modifyValuesByKeys(
    object: object,
    keys: string[],
    modify: (value: string) => string,
) {
    // Clone the object deeply with a customizer function that modifies matching keys
    return cloneDeepWith(object, (value: unknown, key) => {
        if (keys.includes(key as string) && isString(value)) {
            return modify(value);
        }

        return undefined;
    });
}
