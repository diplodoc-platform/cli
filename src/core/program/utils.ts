import type {BaseArgs, BaseConfig} from '~/core/program/types';
import type {BaseProgram} from '~/core/program/index';

import {isAbsolute} from 'node:path';

export function isRelative(path: string | undefined) {
    return /^\.{1,2}\//.test(path || '');
}

export function requireExtension(name: string) {
    if (isRelative(name) || isAbsolute(name)) {
        return require(name);
    }

    const local = (name: string) => {
        if (name.startsWith('@')) {
            return undefined;
        }

        return '@diplodoc/cli/' + name;
    };

    const scoped = (name: string, fixed = false) => {
        if (name.startsWith('@')) {
            return undefined;
        }

        return '@diplodoc/' + name + (fixed ? '-extension' : '');
    };

    const precise = (name: string, fixed = false) => {
        return name + (fixed ? '-extension' : '');
    };

    const variants = [
        local(name),
        scoped(name),
        scoped(name, true),
        precise(name),
        precise(name, true),
    ].filter(Boolean);

    while (variants.length) {
        const variant = variants.shift() as string;
        try {
            return require(variant);
        } catch {}
    }

    throw new Error(`Unable to resolve '${name}' extension`);
}

export function isProgram<TConfig extends BaseConfig, TArgs extends BaseArgs>(
    module: unknown,
): module is BaseProgram<TConfig, TArgs> {
    return Boolean(module && typeof (module as BaseProgram).init === 'function');
}
