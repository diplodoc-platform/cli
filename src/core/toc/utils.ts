import type {TextFilter} from './types';

import {evaluate} from '@diplodoc/liquid';

export function isRelative(path: AnyPath): path is RelativePath {
    return /^\.{1,2}\//.test(path) || !/^(\w{0,7}:)?\/\//.test(path);
}

export function getFirstValuable<T>(
    items: TextFilter[] | string,
    vars: Hash,
    fallback?: T,
): T | undefined {
    if (typeof items === 'string') {
        items = [{text: items, when: true}];
    }

    if (!Array.isArray(items)) {
        items = [];
    }

    for (const item of items) {
        let {when = true} = item;
        delete item.when;

        if (typeof when === 'string') {
            when = Boolean(evaluate(when, vars));
        }

        if (when) {
            return item.text as T;
        }
    }

    return fallback;
}
