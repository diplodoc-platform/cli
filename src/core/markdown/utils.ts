import type {AssetInfo, Location} from './types';

import {parseLocalUrl} from '~/core/utils';

type AssetModifier = '!' | '@' | '';

const modifiers = {'!': 'image', '@': 'video', '': 'link'} as const;

export function findLink(content: string): string | undefined {
    const rx = /]\(\s*/g;
    const match = rx.exec(content);
    if (!match) {
        return undefined;
    }

    const link = parseLinkDestination(content, rx.lastIndex);

    if (link === null || link.match(/[${}]/)) {
        return undefined;
    }

    rx.lastIndex += link.length + 1;

    return link;
}

export function findLinksInfo(content: string): AssetInfo[] {
    const links = find(/]\(\s*/g, content, (match, rx) => {
        const link = parseLinkDestination(content, rx.lastIndex);
        const title = parseLinkTitle(content, rx.lastIndex - match[0].length);

        // TODO: add more precise filter for unix compatible paths
        if (link === null || title === null || link.match(/[${}]/)) {
            return undefined;
        }

        const parsed = parseLocalUrl(link);
        if (!parsed) {
            return undefined;
        }

        const modifier = content[
            rx.lastIndex - match[0].length - title.length - 2
        ] as AssetModifier;
        const type = modifiers[modifier] || 'link';

        rx.lastIndex += link.length + 1;

        return {
            ...parsed,
            type,
            subtype: type === 'image' ? 'image' : null,
            title,
            autotitle: type === 'link' && (!title || title === '{#T}'),
        };
    });

    const referenceLinks = find(/]\[\s*/g, content, (match, rx) => {
        let link = parseLinkDestination(content, rx.lastIndex, ['[', ']']);
        let title = parseLinkTitle(content, rx.lastIndex - match[0].length);

        // TODO: add more precise filter for unix compatible paths
        if ((link === null && title === null) || link?.match(/[${}]/)) {
            return undefined;
        }
        title = title || '';
        link = link || title;

        const parsed = parseLocalUrl(link);
        if (!parsed) {
            return undefined;
        }

        const modifier = content[
            rx.lastIndex - match[0].length - title.length - 2
        ] as AssetModifier;
        const type = modifiers[modifier] || 'link';

        rx.lastIndex += link.length + 1;

        return {
            ...parsed,
            type,
            subtype: 'reference',
            code: link,
            title,
            autotitle: type === 'link' && (!title || title === '{#T}'),
        };
    });

    return [...links, ...referenceLinks];
}

export function findDefs(content: string): AssetInfo[] {
    return find(/^\s*\[(.*?)]:\s*([^\s]+)/gm, content, (match) => {
        const parsed = parseLocalUrl(match[2]);
        if (!parsed) {
            return undefined;
        }

        return {
            ...parsed,
            type: 'def',
            code: match[1] || undefined,
            title: '',
            autotitle: true,
        };
    });
}

function find(
    matcher: RegExp,
    content: string,
    map: (match: RegExpMatchArray, rx: RegExp) => Omit<AssetInfo, 'location'> | void,
): AssetInfo[] {
    const links: AssetInfo[] = [];

    let match;
    // eslint-disable-next-line no-cond-assign
    while ((match = matcher.exec(content))) {
        const result = map(match, matcher);
        if (!result) {
            continue;
        }

        if (result) {
            links.push({
                ...result,
                location: [match.index, matcher.lastIndex],
            });
        }
    }

    return links;
}

function parseLinkDestination(str: string, start: number, symbols = ['(', ')']) {
    const max = str.length;
    const [symbolStart, symbolEnd] = symbols.map((symbol) => symbol.charCodeAt(0));

    let code,
        level = 0,
        pos = start;

    if (str.charCodeAt(pos) === 0x3c /* < */) {
        return parseSimpleLink(str, start + 1);
    }

    while (pos < max) {
        code = str.charCodeAt(pos);

        if (code === 0x20) {
            break;
        }

        // ascii control characters
        if (code < 0x20 || code === 0x7f) {
            break;
        }

        if (code === 0x5c /* \ */ && pos + 1 < max) {
            if (str.charCodeAt(pos + 1) === 0x20) {
                break;
            }
            pos += 2;
            continue;
        }

        if (code === symbolStart /* ( or ] */) {
            level++;
            if (level > 32) {
                return null;
            }
        }

        if (code === symbolEnd /* ) or ] */) {
            if (level === 0) {
                break;
            }
            level--;
        }

        pos++;
    }

    if (start === pos) {
        return null;
    }
    if (level !== 0) {
        return null;
    }

    return str.slice(start, pos);
}

function parseLinkTitle(str: string, start: number) {
    let code,
        prev,
        level = 0,
        pos = start;

    while (pos >= 0) {
        code = str.charCodeAt(pos);
        prev = str.charCodeAt(pos - 1);

        // ascii control characters
        if (code < 0x20 || code === 0x7f) {
            break;
        }

        if (prev === 0x5c /* \ */) {
            pos -= 2;
            continue;
        }

        if (code === 93 /* ] */) {
            level++;
            if (level > 32) {
                return null;
            }
        }

        if (code === 91 /* [ */) {
            if (level === 1) {
                break;
            }
            level--;
        }

        pos--;
    }

    return str.slice(pos + 1, start);
}

function parseSimpleLink(str: string, start: number) {
    const max = str.length;

    let code,
        pos = start;

    while (pos < max) {
        code = str.charCodeAt(pos);
        if (code === 0x0a /* \n */) {
            return null;
        }
        if (code === 0x3c /* < */) {
            return null;
        }
        if (code === 0x3e /* > */) {
            return str.slice(start, pos);
        }
        if (code === 0x5c /* \ */ && pos + 1 < max) {
            pos += 2;
            continue;
        }

        pos++;
    }

    // no closing '>'
    return null;
}

export function filterRanges<T extends {location: Location}>(
    excludes: Location[],
    infos: T[],
): T[] {
    const contains = (exclude: Location, point: Location) => {
        return (
            (exclude[1] >= point[0] && exclude[1] <= point[1]) ||
            (exclude[0] >= point[0] && exclude[0] < point[1]) ||
            (exclude[0] < point[0] && exclude[1] >= point[1])
        );
    };

    return infos.filter((item) => {
        return !excludes.some((exclude) => contains(exclude, item.location));
    });
}

export function parseHeading(content: string) {
    const anchors = [];
    const commonHeading = content.match(/^#+/);
    const alternateHeading = content[content.length - 1];
    const alternaleLevels = ['-', '='];
    const level = commonHeading
        ? commonHeading[0].length
        : alternaleLevels.indexOf(alternateHeading) + 1;

    if (commonHeading) {
        content = content.replace(/^#+\s*/, '');
    } else {
        content = content.replace(/\n[-=]+$/, '');
    }

    const ANCHOR = /{(#[^}]+)}/g;

    let match;
    // eslint-disable-next-line no-cond-assign
    while ((match = ANCHOR.exec(content))) {
        anchors.push(match[1]);
        content = content.replace(match[0], '');
        ANCHOR.lastIndex -= match[0].length;
    }

    const title = content.trim();

    return {anchors, title, level};
}
