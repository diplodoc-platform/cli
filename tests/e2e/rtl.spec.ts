import {getTestPaths, runYfmDocs, compareDirectories} from '../utils';

const generateMapTestTemplate = (testTitle: string, testRootPath: string, {md2md = true, md2html = true, args = '--allow-custom-resources'}) => {
    test(testTitle, () => {
        const {inputPath, outputPath} = getTestPaths(testRootPath);
        runYfmDocs(inputPath, outputPath, {md2md, md2html, args});
        compareDirectories(outputPath);
    });
}

describe('Generate html document with correct lang and dir attributes. Load correct bundles.', () => {
    generateMapTestTemplate('documentation with rtl and ltr langs', 'mocks/rtl/multidirectional-languages', {md2html: true, md2md: false})

    generateMapTestTemplate('documentation with only one rtl lang', 'mocks/rtl/rtl-language', {md2html: true, md2md: false})
});
