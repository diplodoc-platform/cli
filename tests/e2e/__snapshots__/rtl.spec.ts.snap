// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate html document with correct lang and dir attributes. Load correct bundles. documentation with only one rtl lang 1`] = `
"[
  "_bundle/search-async-0",
  "_bundle/app-css-1",
  "_bundle/app-js-1",
  "_bundle/search-css-2",
  "_bundle/search-js-0",
  "_bundle/search-js-1",
  "_bundle/search-css-0",
  "_bundle/search-js-2",
  "_bundle/search-css-1",
  "index.html",
  "page.html"
]"
`;

exports[`Generate html document with correct lang and dir attributes. Load correct bundles. documentation with only one rtl lang 2`] = `
<!DOCTYPE html>
<html lang="he"
      dir="rtl"
>
  <head>
    <meta charset="utf-8">
    <base href="./">
    <meta name="viewport"
          content="width=device-width, initial-scale=1.0"
    >
    <title>
      Documentation
    </title>
    <style type="text/css">
      body {
                        height: 100vh;
                    }
    </style>
    <link type="text/css"
          rel="stylesheet"
          href="_bundle/search-css-1"
    >
    <link type="text/css"
          rel="stylesheet"
          href="_bundle/search-css-2"
    >
  </head>
  <body class="g-root g-root_theme_light">
    <div id="root">
    </div>
    <script type="application/javascript">
      window.STATIC_CONTENT = false
                   window.__DATA__ = {"data":{"leading":true,"toc":{"title":"Documentation","href":"index.html","items":[{"name":"Documentation","href":"page.html","id":"Documentation-RANDOM"}],"base":".","deepBase":0},"data":{"title":"Documentation","links":[{"title":"Documentation","href":"page.html"}]},"meta":{"metadata":[],"style":[],"script":[]}},"router":{"pathname":"index.html","depth":1},"lang":"he","langs":["he"]};
    </script>
    <script type="application/javascript"
            src="_bundle/search-js-0"
    >
    </script>
    <script type="application/javascript"
            src="_bundle/app-js-1"
    >
    </script>
    <script type="application/javascript"
            src="_bundle/search-js-2"
    >
    </script>
  </body>
</html>
`;

exports[`Generate html document with correct lang and dir attributes. Load correct bundles. documentation with only one rtl lang 3`] = `
<!DOCTYPE html>
<html lang="he"
      dir="rtl"
>
  <head>
    <meta charset="utf-8">
    <base href="./">
    <meta name="viewport"
          content="width=device-width, initial-scale=1.0"
    >
    <title>
      Documentation
    </title>
    <style type="text/css">
      body {
                        height: 100vh;
                    }
    </style>
    <link type="text/css"
          rel="stylesheet"
          href="_bundle/search-css-1"
    >
    <link type="text/css"
          rel="stylesheet"
          href="_bundle/search-css-2"
    >
  </head>
  <body class="g-root g-root_theme_light">
    <div id="root">
    </div>
    <script type="application/javascript">
      window.STATIC_CONTENT = false
                   window.__DATA__ = {"data":{"leading":false,"toc":{"title":"Documentation","href":"index.html","items":[{"name":"Documentation","href":"page.html","id":"Documentation-RANDOM"}],"base":".","deepBase":0},"meta":{"metadata":[],"style":[],"script":[]},"assets":[],"headings":[],"title":"","includes":[],"html":"
      <p>תיעוד
      </p>/n"},"router":{"pathname":"page.html","depth":1},"lang":"he","langs":["he"]};
    </script>
    <script type="application/javascript"
            src="_bundle/search-js-0"
    >
    </script>
    <script type="application/javascript"
            src="_bundle/app-js-1"
    >
    </script>
    <script type="application/javascript"
            src="_bundle/search-js-2"
    >
    </script>
  </body>
</html>
`;

exports[`Generate html document with correct lang and dir attributes. Load correct bundles. documentation with rtl and ltr langs 1`] = `
"[
  "_bundle/search-async-0",
  "_bundle/app-css-1",
  "_bundle/app-js-1",
  "_bundle/search-css-2",
  "_bundle/search-js-0",
  "_bundle/search-js-1",
  "_bundle/search-css-0",
  "_bundle/search-js-2",
  "_bundle/search-css-1",
  "ar/index.html",
  "ar/page.html",
  "en/index.html",
  "en/page.html",
  "index.html"
]"
`;

exports[`Generate html document with correct lang and dir attributes. Load correct bundles. documentation with rtl and ltr langs 2`] = `
<!DOCTYPE html>
<html lang="ar"
      dir="rtl"
>
  <head>
    <meta charset="utf-8">
    <base href="./">
    <meta name="viewport"
          content="width=device-width, initial-scale=1.0"
    >
    <title>
      Documentation
    </title>
    <style type="text/css">
      body {
                        height: 100vh;
                    }
    </style>
    <link type="text/css"
          rel="stylesheet"
          href="../_bundle/search-css-1"
    >
    <link type="text/css"
          rel="stylesheet"
          href="../_bundle/search-css-2"
    >
  </head>
  <body class="g-root g-root_theme_light">
    <div id="root">
    </div>
    <script type="application/javascript">
      window.STATIC_CONTENT = false
                   window.__DATA__ = {"data":{"leading":true,"toc":{"title":"Documentation","href":"index.html","items":[{"name":"Documentation","href":"page.html","id":"Documentation-RANDOM"}],"base":"ar","deepBase":1},"data":{"title":"Documentation","links":[{"title":"Documentation","href":"page.html"}]},"meta":{"metadata":[],"style":[],"script":[]}},"router":{"pathname":"index.html","depth":2},"lang":"ar","langs":["ar","en"]};
    </script>
    <script type="application/javascript"
            src="../_bundle/search-js-0"
    >
    </script>
    <script type="application/javascript"
            src="../_bundle/app-js-1"
    >
    </script>
    <script type="application/javascript"
            src="../_bundle/search-js-2"
    >
    </script>
  </body>
</html>
`;

exports[`Generate html document with correct lang and dir attributes. Load correct bundles. documentation with rtl and ltr langs 3`] = `
<!DOCTYPE html>
<html lang="ar"
      dir="rtl"
>
  <head>
    <meta charset="utf-8">
    <base href="./">
    <meta name="viewport"
          content="width=device-width, initial-scale=1.0"
    >
    <title>
      Documentation
    </title>
    <style type="text/css">
      body {
                        height: 100vh;
                    }
    </style>
    <link type="text/css"
          rel="stylesheet"
          href="../_bundle/search-css-1"
    >
    <link type="text/css"
          rel="stylesheet"
          href="../_bundle/search-css-2"
    >
  </head>
  <body class="g-root g-root_theme_light">
    <div id="root">
    </div>
    <script type="application/javascript">
      window.STATIC_CONTENT = false
                   window.__DATA__ = {"data":{"leading":false,"toc":{"title":"Documentation","href":"index.html","items":[{"name":"Documentation","href":"page.html","id":"Documentation-RANDOM"}],"base":"ar","deepBase":1},"meta":{"metadata":[],"style":[],"script":[]},"assets":[],"headings":[],"title":"","includes":[],"html":"
      <p>توثيق
      </p>/n"},"router":{"pathname":"page.html","depth":2},"lang":"ar","langs":["ar","en"]};
    </script>
    <script type="application/javascript"
            src="../_bundle/search-js-0"
    >
    </script>
    <script type="application/javascript"
            src="../_bundle/app-js-1"
    >
    </script>
    <script type="application/javascript"
            src="../_bundle/search-js-2"
    >
    </script>
  </body>
</html>
`;

exports[`Generate html document with correct lang and dir attributes. Load correct bundles. documentation with rtl and ltr langs 4`] = `
<!DOCTYPE html>
<html lang="en"
      dir="ltr"
>
  <head>
    <meta charset="utf-8">
    <base href="./">
    <meta name="title"
          content="Documentation"
    >
    <meta name="noIndex"
          content="true"
    >
    <meta name="viewport"
          content="width=device-width, initial-scale=1.0"
    >
    <title>
      Documentation | Documentation
    </title>
    <style type="text/css">
      body {
                        height: 100vh;
                    }
    </style>
    <link type="text/css"
          rel="stylesheet"
          href="../_bundle/search-css-0"
    >
    <link type="text/css"
          rel="stylesheet"
          href="../_bundle/app-css-1"
    >
  </head>
  <body class="g-root g-root_theme_light">
    <div id="root">
    </div>
    <script type="application/javascript">
      window.STATIC_CONTENT = false
                   window.__DATA__ = {"data":{"leading":true,"toc":{"title":"Documentation","href":"index.html","items":[{"name":"Documentation","href":"page.html","id":"Documentation-RANDOM"}],"base":"en","deepBase":1},"data":{"title":"Documentation","description":"","meta":{"title":"Documentation","noIndex":true,"metadata":[],"style":[],"script":[]},"links":[{"title":"Getting started with Documentation","description":"This guide will show you the basics of working with Documentation","href":"page.html"}]},"meta":{"title":"Documentation","noIndex":true,"metadata":[],"style":[],"script":[]}},"router":{"pathname":"index.html","depth":2},"lang":"en","langs":["ar","en"]};
    </script>
    <script type="application/javascript"
            src="../_bundle/search-js-0"
    >
    </script>
    <script type="application/javascript"
            src="../_bundle/app-js-1"
    >
    </script>
    <script type="application/javascript"
            src="../_bundle/search-js-2"
    >
    </script>
  </body>
</html>
`;

exports[`Generate html document with correct lang and dir attributes. Load correct bundles. documentation with rtl and ltr langs 5`] = `
<!DOCTYPE html>
<html lang="en"
      dir="ltr"
>
  <head>
    <meta charset="utf-8">
    <base href="./">
    <meta name="yfm"
          content="builder"
    >
    <meta name="viewport"
          content="width=device-width, initial-scale=1.0"
    >
    <title>
      Documentation
    </title>
    <style type="text/css">
      body {
                        height: 100vh;
                    }
    </style>
    <link type="text/css"
          rel="stylesheet"
          href="../_bundle/search-css-0"
    >
    <link type="text/css"
          rel="stylesheet"
          href="../_bundle/app-css-1"
    >
  </head>
  <body class="g-root g-root_theme_light">
    <div id="root">
    </div>
    <script type="application/javascript">
      window.STATIC_CONTENT = false
                   window.__DATA__ = {"data":{"leading":false,"toc":{"title":"Documentation","href":"index.html","items":[{"name":"Documentation","href":"page.html","id":"Documentation-RANDOM"}],"base":"en","deepBase":1},"meta":{"metadata":[{"name":"yfm","content":"builder"}],"style":[],"script":[]},"assets":[],"headings":[],"title":"","includes":[],"html":"
      <p>Lorem
      </p>/n"},"router":{"pathname":"page.html","depth":2},"lang":"en","langs":["ar","en"]};
    </script>
    <script type="application/javascript"
            src="../_bundle/search-js-0"
    >
    </script>
    <script type="application/javascript"
            src="../_bundle/app-js-1"
    >
    </script>
    <script type="application/javascript"
            src="../_bundle/search-js-2"
    >
    </script>
  </body>
</html>
`;

exports[`Generate html document with correct lang and dir attributes. Load correct bundles. documentation with rtl and ltr langs 6`] = `
<!DOCTYPE html>
<html lang="ar"
      dir="rtl"
>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="refresh"
          content="0; url=./ar/index.html"
    >
    <meta name="viewport"
          content="width=device-width, initial-scale=1.0"
    >
    <title>
      Redirect
    </title>
    <style type="text/css">
      body {
                        height: 100vh;
                    }
    </style>
    <link type="text/css"
          rel="stylesheet"
          href="_bundle/search-css-1"
    >
    <link type="text/css"
          rel="stylesheet"
          href="_bundle/search-css-2"
    >
    <script type="text/javascript">
      window.location.replace("./ar/index.html");
    </script>
  </head>
  <body class="g-root g-root_theme_light">
    If you are not redirected automatically, follow this
    <a href="./ar/index.html">
      link to example
    </a>
    .
  </body>
</html>
`;
